### DISCUSSION TITLE
On hashability and Substitutability of MultiIndexLocation vs. IndexLocation

### DISCUSSION THREAD

**Discussion by drewj-tp** [terrapower/armi#1930] [General]
I saw this comment and I wonder when it's time to address the things the comment indicates should be addressed

https://github.com/terrapower/armi/blob/b436537edfed991e6ff560a1d1b2d4b018507db1/armi/reactor/grids/locations.py#L369-L371

Substitutability meaning you can't use a multiindexlocation in the same way you'd use an indexlocation, I presume. Should the multi be a subclass of the single index?

**Comment by john-science** [comment]
That's such a strange comment. Obviously, we can just implement our own `def __hash__(self)` method in any class we want. Why would that be impossible here?

**Reply by drewj-tp** [reply]
Speculating, but I think the comment stems from the mutability of the `MultiIndexLocation`

https://github.com/terrapower/armi/blob/b436537edfed991e6ff560a1d1b2d4b018507db1/armi/reactor/grids/locations.py#L422-L429

If you made a MultiIndexLocation hashable based on the hash-values of its contained locations, the hash value would change as you add/remove more locations.

Trivially you could hash based off the id of the object. But then you can't substitute in another multi index location that has similar data

**Reply by john-science** [reply]
Sure, you have to DECIDE if the hash of a `MultiIndexLocation` depends on the full set of locations currently in it.  But the hash of an object (nearly?) always depends on the state of that object. And if you want to decide otherwise, great.

Still, it seems more solvable than the comment suggests. There's a decision to make, is all.
